### ISSUE TITLE
[Bug]: Store the Notebook Image Name

### ISSUE THREAD

**Issue by andrewballantyne** [opendatahub-io/odh-dashboard#1370]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

In the table of workbenches, we explain the notebook image in play. This is a hard reference to images we have in play at this point in time.

However, when we keep moving n & n+1 images forward... older notebooks lose that reference and fall into "Unknown" image name. This is not ideal for upgrades and on-going improvements for the Dashboard to cause "loss" of image.

### Expected Behavior

Store the image name for when we lose it, it can still represent the image in the table -- but it won't be able to find it in the list of edit.

Maybe UX can let us know if we should identify it in some way or just make it seamless.

### Steps To Reproduce

1. Create a new image
2. Use it in a workbench creation
3. Delete it

See the table as `Unknown` -- this is effectively the effort behind upgrades.

### Open Data Hub Version

```yml
Dashboard: v2.11.0
```

**Comment by andrewballantyne** [comment]
@kywalker-rh / @vconzola Thoughts on if we should denote "images that are old and untracked"? Essentially when we keep up-versioning images, the oldest ones fall off and the table represents them as Unknown -- but we should probably at least keep the name, but would we want to perhaps show "(deprecated)" or something like that beside it? Essentially the image as the name has either been deleted or the tag has moved on and you can't recreate this workbench anymore.

Or do you guys just feel we shouldn't do anything and just leave it as-is. On Edit, we won't be able to select the image because it does not exist anymore.

**Comment by vconzola** [comment]
@andrewballantyne What happens to existing workbenches that are using "old" images? Do they automatically get updated to use the newer version of the image?

**Comment by andrewballantyne** [comment]
@vconzola Nope, nothing gets automatically updated when it is created by the user; such as Workbenches. The items tracked by the Operator (our OOTB images for instance) are the only ones that get "auto updated". 

The way we render the name is a life representation to what is on the cluster (eg. the Workbench stores "I use image x, on tag y" then we ask image "x:y", what is your name and render that) -- which when we do the n / n+1 upgrades, eventually falls off the cluster. Anything using it works fine, but you cannot create more with that image; you'd need to use a later one.

**Comment by shalberd** [comment]
@vconzola @andrewballantyne 

You have to be careful not to confuse images and imagestream name plus tag. I see what you mean by the name getting deleted (i.e. the recent tag name changes of imagestreams from v.3.9 to 2023.1 or so) or tag moving on (same tag of an image behind a recommended image with -weekly image tag (not imagestream tag) ) pointing to an old digest that is not present in the openshift internal repo anymore).  

My recent proposal PR-800, initially with regards to image-field values in digest format, of storing the initial imgestream name and tag value in the env var JUPYTER_IMAGE might at least help keep the imagestream name and tag persistent, but true, you will loose access to older -weekly image versions as the -weekly tag in external quay.io moves on currently.

Pointing to an sha256 digest, which is also part of my proposal, could alleviate that problem by not pointing to the weekly tag in the internal openshift repo, which always points to the latest digest, but to the digest, which is kept in cache for a while ( last three digests values). ([keepTagRevisions](https://docs.openshift.com/container-platform/4.10/applications/pruning-objects.html#pruning-images_pruning-objects))

**Comment by kywalker-rh** [comment]
> but would we want to perhaps show "(deprecated)" or something like that beside it?

@andrewballantyne Yes, I think that's a good idea. Deprecated even makes sense to me, although I'm open to suggestions if @kaedward thinks there is a better word for an image that is no longer supported.

**Comment by kaedward** [comment]
@andrewballantyne @kywalker-rh In RHOSAK, we used "Degraded" to refer to an instance that is experiencing a non-terminal error. It sounds like if this image is no longer supported, though, that "deprecated" might be more accurate. So thumbs up from me :)

**Comment by andrewballantyne** [comment]
Thanks for the UX Input @kywalker-rh & @kaedward let me try to figure out how to make use of it.

Okay, so I followed most of what you mentioned @shalberd -- here is my suggestion, let me know if I'm over complicating this.

1. This solution will not apply to Jupyter tile as they don't quite refer to things the same way, and we have other objectives to improve that
2. On Workbench create/update, we add the annotation `opendatahub.io/image-display-name` and at that time, determine based on the image selected, what we will use
3. We update our Workbench table to use that
4. On render of the Workbench table rows, we will attempt to resolve the image TAG (😉) -- if we can't find it, we will suffix the display name with `(deprecated)` -- eg. `Pytorch (deprecated)` -- might need a use case or two done to make sure it looks alright to the UX folks

That way it's better handling the "no reference" as well as the on-going "there is a reference" scenarios. On Edit of the workbench, we just need to make sure we don't "break" -- the user will need to reselect an image anyways.

**Comment by shalberd** [comment]
Hi @andrewballantyne 

>This solution will not apply to Jupyter tile

I am not sure what that means, it seemed to me so far you had similar logic for looking up imagestream name and tag from the container image field in backend and frontend. My main point there was that it is impossible to get the imagestream tag if ever the container image field value has sha256 digest notation, which would happen if we use the image change trigger plugin.

>On Workbench create/update, we add the annotation opendatahub.io/image-display-name and at that time, determine based on the image selected, what we will use

I assume that the the assembleNotebook / updateNotebook part, where the notebook CR yaml is constructed. If yes, then there already is an annotation called `notebooks.opendatahub.io/last-image-selection` that contains the last selected imagestream name and tag, e.g. jupyter-datascience:2023.1. up to now, the variable imageSelection that contains that string is also used for that annotation, in [function assembleNotebook](https://github.com/opendatahub-io/odh-dashboard/blob/main/frontend/src/api/k8s/notebooks.ts#L50).

You could either use the value in that annotation or, if my PR 800 with the new mechanism for container image field lookup from imagestream name and tag evers makes it into production, from env variable JUPYTER_IMAGE, that would also then have imageSelection.

Fully ACK on point 4.

**Comment by shalberd** [comment]
@atheo89 The problem is in that case, the imagestream tag changed, similar to [here](https://github.com/opendatahub-io/odh-manifests/commit/56a6763ef0afffeeac18dab414c96897009acd65#diff-eec1054bc84f334a3bcb37d2a71e8f2242c20ab4819f96a2623adcdde17cdc88) .... there is no way to get the related info when an imagestream yaml was changed. This is purely a matter of fast-changing master-branch or comparable dev changes in my view, not a conceptual problem.

image-registry.openshift-image-registry.svc:5000/redhat-ods-applications/ s2i-minimal-notebook:1.2 is cached, not pruned yet in the internal registry. That explains why the pod container is still running. But I am almost 100% certain that in this scenario, the imagestream tag changed, or the imagestream itself was deleted ...

**Comment by shalberd** [comment]
The current tag-based notation is not suitable for on-prem and mirrored environments anyways, because it does not use sha256 notation. In adddition, it also always references the internal openshift registry, even when no internal openshift registry is present. That's why my proposed change in [PR-800](https://github.com/opendatahub-io/odh-dashboard/pull/800) But even that PR would not have solved the problem of image metadata getting lost when either the imagestream metadata.name changes or the tags in an imagestream change.

**Comment by atheo89** [comment]
@shalberd oh, sorry I corrected my comment above, I experimented with 1.2-new tag not 1.2. I tested with that changed tag then I deleted it and checked again, and all the above mentions shows the tag that the image has been created a d it seems that it persists.

**Comment by shalberd** [comment]
And even if you continuously checked an imagestream tag's status conditions

https://docs.okd.io/4.11/rest_api/image_apis/imagestream-image-openshift-io-v1.html#status-tags-conditions

i.e. transitionTime for event ImportSuccess via the notebook controller, even then you'd have to decide what to update the notebook with:

1.2-new it was created with, now it is 1.2 ... but how would you know that it now needs to refer to tag 1.2? Might just as well be another tag 1.3 or 2023-stable or whatever in the imagestream tags array and tag 1.2 is not there anymore in the imagestream yaml ...

**Comment by shalberd** [comment]
I use a different function to get all the image details, from [env JUPYTER_IMAGE containing now the value
imagestreamname:imagestreamtag](https://github.com/opendatahub-io/odh-dashboard/pull/800/files#diff-677151b7028c8a28b9d290ca5f530c46c4e73e540341e7ed1de0b6472c98b5c9), but that is conceptually the same as  image-registry.openshift-image-registry.svc:5000/redhat-ods-applications/ imagestreamname:imagestreamtag in how dashboard parses image information itself. Meaning that if the old image-tag is present in the container field, or in my case, the JUPYTER_IMAGE env var, you still don't get the details of the tag for the imagestream ...

**Comment by andrewballantyne** [comment]
I don't think it was addressed in the comments here... but `opendatahub.io/image-display-name` is a Notebook annotation. It's for the Dashboard to handle the name. If a Notebook is created outside of the Dashboard, and the image reference is no longer available, it will not reflect... hence `Unknown` is accurate.

We could theoretically try a fallback operation here and look at the image-stream to try and figure out a display name. Most likely case here is the tag slips off the cluster, not the whole image stream. But in the BYON use-case, you could lose the entire image-stream -- of which all data is lost.

We can do this through a few stages and layers. To make this one simple, we could do my `image-display-name` idea and just have metadata we can pull upon. We can improve the fallover point for the metadata by using the image-stream and ignoring the current tag (if it's outdated). That'll just improve the `(deprecated)` use-cases.

**Comment by shalberd** [comment]
@harshad16 Yes, I just tried the same thing with my PR-800 image change trigger annotation. 

First referenced an image with an imagestream tag. Notebook starts up. Then changed that imagestream tag in the imagestream spec to a new value. 
Notebook keeps running because the old tag, though not in the imagestream spec.tags anymore, is _still in the imagestream status section, containing the image name and same digest._ 

But not more than that. So, even there, no more details on the existing tag in odh dashboard. Makes sense because the spec.tags[x] with all the details is not present for the old tag. 

![Bildschirmfoto 2023-06-30 um 11 10 48](https://github.com/opendatahub-io/odh-dashboard/assets/21118431/316c6d33-2b47-4259-b290-e897343e2857)


I agree with @andrewballantyne that in such a case, setting the image name to Unknown in dashboard image details is a good idea. You just won't have the additional info ([packages, software and so on](https://github.com/opendatahub-io/odh-manifests/blob/master/notebook-images/base/jupyter-datascience-notebook-imagestream.yaml#L31)) anymore that was associated with the old imagestream tag.

**Comment by shalberd** [comment]
if my change in PR-800, putting imagestream name and tag in env var JUPYTER_IMAGE  i.e. jupyter-datascience:py3.9, were approved, then we would not need a new annotation like image-display-name in the notebook CR, then we could just use annotation  notebooks.opendatahub.io/last-image-selection and fill it with the value from the `[imagestream annotation ](https://github.com/opendatahub-io/odh-manifests/blob/master/notebook-images/base/jupyter-datascience-notebook-imagestream.yaml#L8)` Meaning in the code, [lastSelectedImage](last-image-selection) could come from container env var JUPYTER_IMAGE in future.

**Comment by shalberd** [comment]
@VaishnaviHire @harshad16 @andrewballantyne on second thought, though, this here is an enhancement for tracking imagestream metadata on the GUI. It'd be nice if we could finally pull in the changes of [PR-800](https://github.com/opendatahub-io/odh-dashboard/pull/800) dashboard and[ PR-72](https://github.com/opendatahub-io/kubeflow/pull/72) KF notebook controller.

**Comment by shalberd** [comment]
@andrewballantyne 
> To make this one simple, we could do my image-display-name idea and just have metadata we can pull upon. We can improve the fallover point for the metadata by using the image-stream and ignoring the current tag (if it's outdated). That'll just improve the (deprecated) use-case

Full ACK, good idea

**Comment by harshad16** [comment]
I finally caught up with the thread.

With the imagestream tag deletion, the imagestream reference would be pruned from the internal registry with the garbage-collection runs, though as the workbench pod was already running, as the container was started, it would
only start to fail if the notebook somehow gets restarted or re-kicked.
Which I feel should be the standard, as it is on the user who deleted the tag, and still kept a workbench referenced
to the deleted tag.

> To make this one simple, we could do my image-display-name idea and just have metadata we can pull upon. We can > improve the fallover point for the metadata by using the image-stream and ignoring the current tag (if it's outdated). > > That'll just improve the (deprecated) use-cas

@andrewballantyne from this comment, I take the action items as, including image-stream annotation `image-display-name` on the OOTB notebook image, if not present already. 
For the imagestream from BYON, it is done from the dashboard code, so it could be handled from there.

Please let us know, if any further assistance is needed.

**Comment by harshad16** [comment]
> on second thought, though, this here is an enhancement for tracking imagestream metadata on the GUI. Let's proceed step by step.
> It'd be nice if we could finally pull in the changes of https://github.com/opendatahub-io/odh-dashboard/pull/800 
> dashboard andhttps://github.com/opendatahub-io/kubeflow/pull/72 KF notebook controller.

@shalberd ack, i think as you described this seems to be a great feature to have and is responding a lot of the concerns around the imagestream.
Will include it in the KF controller.

**Comment by shalberd** [comment]
> With the imagestream tag deletion, the imagestream reference would be pruned from the internal registry

Eventually, yes. We could think about setting notebook container imagePullPolicy to IfNotPresent, which is a sensible change from Always, but at some point, even the cache on then Openshift nodes will not be there anymore.

**Comment by shalberd** [comment]
> from this comment, I take the action items as, including image-stream annotation image-display-name on the OOTB notebook image, if not present already.

there is an imagestream annotation for the imagestream / image called [opendatahub.io/notebook-image-name](https://github.com/opendatahub-io/odh-manifests/blob/master/notebook-images/base/jupyter-datascience-notebook-imagestream.yaml#L8C5-L8C39) I am not familiar with the BYON code for assembling imagestream, but that could be useful, right? You could add a notebook annotation `opendatahub.io/image-display-name based` on that.

**Comment by shalberd** [comment]
@harshad16 @andrewballantyne changed [imagePullPolicy to IfNotPresent i](https://github.com/opendatahub-io/odh-dashboard/pull/800/commits/c50b0bd3b89de1154a7d5612526b5ba8fe106793)n Notebook CR assembly. Also in some mock notebooks, except nvidia-dcgm-mock-exporter file, which uses an image with :latest tag, for which "Always" is needed.

https://github.com/opendatahub-io/odh-dashboard/pull/800/commits/c50b0bd3b89de1154a7d5612526b5ba8fe106793

**Comment by andrewballantyne** [comment]
This is still having friction and I'm bumping the priority so we can get to this sooner vs later.
